File: MATLAB_Prototype/CLTStopping.m
classdef CLTStopping < stoppingCriterion
% Stopping criterion based on the Central Limit Theorem
properties
   discDistAllowed = 'IIDDistribution' %which discrete distributions are supported
   decompTypeAllowed = {'single', 'multi'} %which decomposition types are supported
   inflate = 1.2 %inflation factor
   alpha = 0.01;
end

properties (Dependent)
   quantile
end

methods
   function [obj, dataObj, distribObj] = ...
         stopYet(obj, dataObj, funObj, distribObj)
      if ~numel(dataObj)
         dataObj = meanVarData; %create a new accumulated data object
      end
      switch dataObj.stage
         case 'begin' %initialize
            dataObj.timeStart = tic; %keep track of time
            if ~any(strcmp(obj.discDistAllowed,class(distribObj)))
               error('Stopping criterion not compatible with sampling distribution')
            end
            nf = numel(funObj); %number of functions whose integrals add up to the solution
            distribObj = initStreams(distribObj,nf); %need an IID stream for each function
            dataObj.prevN = zeros(1,nf); %initialize data object
            dataObj.nextN = repmat(obj.nInit,1,nf);
            dataObj.muhat = Inf(1,nf);
            dataObj.sighat = Inf(1,nf);
            dataObj.nSigma = obj.nInit; %use initial samples to estimate standard deviation
            dataObj.costF = zeros(1,nf);
            dataObj.stage = 'sigma'; %compute standard deviation next
         case 'sigma'
            dataObj.prevN = dataObj.nextN; %update place in the sequence
            tempA = sqrt(dataObj.costF); %use cost of function values to decide how to allocate
            tempB = sum(tempA .* dataObj.sighat); %samples for computation of the mean
            nM = ceil((tempB*(obj.quantile*obj.inflate ...
               /max(obj.absTol,dataObj.solution*obj.relTol))^2) ...
               * (dataObj.sighat./sqrt(dataObj.costF)));
            dataObj.nMu = min(max(dataObj.nextN,nM),obj.nMax - dataObj.prevN);
            dataObj.nextN = dataObj.nMu + dataObj.prevN;
            dataObj.stage = 'mu'; %compute sample mean next
         case 'mu'
            dataObj.solution = sum(dataObj.muhat);
            dataObj.nSamplesUsed = dataObj.nextN;
            errBar = (obj.quantile * obj.inflate) * ...
               sqrt(sum(dataObj.sighat.^2/dataObj.nMu));
            dataObj.errorBound = dataObj.solution + errBar*[-1 1];
            dataObj.stage = 'done'; %finished with computation
      end
      dataObj.timeUsed = toc(dataObj.timeStart);
   end
 
   function value = get.quantile(obj)
      value = -norminv(obj.alpha/2);
   end
   
 end
end

File: Python_Prototype/algorithms/CLTStopping.py
''' Originally developed in MATLAB by Fred Hickernell. Translated to python by Sou-Cheng T. Choi and Aleksei Sorokin '''
from time import time

from numpy import zeros, full, inf, array, maximum, minimum, tile, ceil
from scipy.stats import norm

from algorithms.meanVarData import meanVarData
from algorithms.stoppingCriterion import stoppingCriterion


class CLTStopping(stoppingCriterion):
    ''' Stopping criterion based on the Centeral Limit Theorem. '''
    def __init__(self,inflate=1.2,alpha=0.01,absTol=None,relTol=None,nInit=None,nMax=None):
        discDistAllowed = ["IIDDistribution"] # which discrete distributions are supported
        decompTypeAllowed = ["single", "multi"] # which decomposition types are supported
        super().__init__(discDistAllowed,decompTypeAllowed,absTol=absTol,relTol=relTol,nInit=nInit,nMax=nMax)
        self.inflate = inflate  # inflation factor
        self.alpha = alpha # uncertainty level

    def stopYet(self, dataObj=None, funObj=[], distribObj=[]):
        if dataObj==None: dataObj=meanVarData(len(funObj))
        if dataObj.stage == 'begin':  # initialize
            dataObj._timeStart = time()  # keep track of time
            if type(distribObj).__name__ not in self.discDistAllowed:
                raise Exception('Stopping criterion not compatible with sampling distribution')
            nf = len(funObj)
            dataObj.prevN = zeros(nf)  # initialize data object
            dataObj.nextN = tile(self.nInit,nf)
            dataObj.muhat = full(nf,inf)
            dataObj.sighat = full(nf,inf)
            dataObj.nSigma = self.nInit  # use initial samples to estimate standard deviation
            dataObj.costF = zeros(nf)
            dataObj.stage = 'sigma'  # compute standard deviation next
        elif dataObj.stage == 'sigma':
            dataObj.prevN = dataObj.nextN  # update place in the sequence
            tempA = (dataObj.costF)**.5  # use cost of function values to decide how to allocate
            tempB = (tempA * dataObj.sighat).sum(0)  # samples for computation of the mean            
            nM = ceil(tempB*(self.get_quantile()*self.inflate / max(self.absTol, dataObj.solution*self.relTol))**2
                * (dataObj.sighat/dataObj.costF**.5))
            dataObj.nMu = minimum(maximum(dataObj.nextN,nM),self.nMax-dataObj.prevN)
            dataObj.nextN = dataObj.nMu + dataObj.prevN
            dataObj.stage = 'mu'  # compute sample mean next
        elif dataObj.stage == 'mu':
            dataObj.solution = dataObj.muhat.sum(0)
            dataObj.nSamplesUsed = dataObj.nextN
            errBar = self.get_quantile() * self.inflate * (dataObj.sighat**2 / dataObj.nMu).sum(0)**.5
            dataObj.confidInt = dataObj.solution + errBar * array([-1, 1])
            dataObj.stage = 'done'  # finished with computation
        dataObj.timeUsed = time() - dataObj._timeStart
        return dataObj, distribObj
    
    def get_quantile(self):
        # dependent property
        return -norm.ppf(self.alpha / 2)

if __name__ == "__main__":
    # Run Doctests
    import doctest
    x = doctest.testfile("Tests/dt_CLTStopping.py")
    print("\n" + str(x))
